nifi/nifi-nar-bundles/nifi-kite-bundle/nifi-kite-processors/src/main/java/org/apache/nifi/processors/kite/AvroRecordConverter.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.nifi.processors.kite;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Field;
+import org.apache.avro.generic.GenericData.Record;
+import org.apache.avro.generic.IndexedRecord;
+import org.codehaus.jackson.JsonNode;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+/**
+ * Responsible for converting records of one Avro type to another. Supports
+ * syntax like "record.field" to unpack fields and will try to do simple type
+ * conversion.
+ */
+public class AvroRecordConverter {
+	private final Schema inputSchema;
+	private final Schema outputSchema;
+	private final Map<String, String> fieldMapping;
+
+	public AvroRecordConverter(Schema inputSchema, Schema outputSchema,
+			JsonNode fieldMapping) {
+		this.inputSchema = inputSchema;
+		this.outputSchema = outputSchema;
+		this.fieldMapping = getFieldMapping(fieldMapping);
+	}
+
+	/**
+	 * Converts one record to another given a input and output schema plus
+	 * explicit mappings for certain target fields.
+	 * 
+	 * @param record
+	 * @param inputSchema
+	 * @param outputSchema
+	 * @param fieldMapping
+	 * @return
+	 * @throws AvroConversionException
+	 */
+	public Record convert(Record input) throws AvroConversionException {
+		Record result = new Record(outputSchema);
+		for (Field outputField : outputSchema.getFields()) {
+			String inputFieldName = outputField.name();
+			if (fieldMapping.containsKey(outputField.name())) {
+				inputFieldName = fieldMapping.get(outputField.name());
+			}
+
+			List<String> fieldParts = Lists.newArrayList(inputFieldName
+					.split("\\."));
+			Record current = input;
+			while (fieldParts.size() > 1) {
+				// Step into the nested records as far as needed.
+				current = (Record) current.get(fieldParts.remove(0));
+			}
+
+			// Current should now be in the right place to read the record.
+			Field f = getFieldForName(inputFieldName, inputSchema);
+			Object content = getContentForName(input, inputFieldName,
```

Comment:
You might consider combining `getFieldForName` and `getContentForName` to avoid stepping through the schema structure twice per value. You'd have to do the conversion inside the last level though, and pass the schema in.

Also, the Kite `EntityAccessor` has a version of `getContentForName` that caches the structure you could copy.